File: gleim300Code/MistyTest/GleimMenu.swift
//
//  GleimMenuList.swift
//  MistyTest
//
//  Created by Jahresprojekt2017/18 on 08.04.18.
//  Copyright © 2018 Ivo Max Muellner. All rights reserved.
//

import Foundation

/*
 Creates subMenu of the portraits
 */
class GleimMenu{
    
    private var subMenuList = [GleimMenu]()
    private var name = ""
    private var audioFilePath = ""
    private var commandList = [String]()
    private var serialCommand = ""
    private var returnCommandList = [String]()
    private var topic = false
    
    /*constructor*/
    init(name: String, subMenu: Array<GleimMenu>, audioFilePath: String, commandList: Array<String>, serialCommand: String, returnCommandList: Array<String>, topic: Bool){
        self.name=name
        self.subMenuList=subMenu
        self.audioFilePath=audioFilePath
        self.commandList=commandList
        self.serialCommand=serialCommand
        self.returnCommandList=returnCommandList
        self.topic=topic
    }
    
    init(name: String){
        self.name=name
    }
    
    init(){
        
    }
    
    /*
     Sets submenu for each portrait
     */
    public static func getParentMenu(childOfParent :GleimMenu, searchMenu: GleimMenu) -> GleimMenu {
        for subMenu in searchMenu.getSubMenuList() {
            if childOfParent.getName() == subMenu.getName() {
                return searchMenu
            }
        }
        for subMenu in searchMenu.getSubMenuList() {
            let potentialParent = getParentMenu(childOfParent: childOfParent, searchMenu: subMenu) //recursive
            if potentialParent.getName() != "" {
                return potentialParent
            }
        }
        
        return GleimMenu()
    }
    
    public func getSubMenuList() -> Array<GleimMenu> {
        return subMenuList
    }
    
    public func getName() -> String {
        return name
    }
    
    public func getAudioFilePath() -> String {
        return audioFilePath
    }
    
    public func getOwnCommandList() -> Array<String> {
        return commandList
    }
    
    public func getSubMenuCommandList() -> Array<String> {
        var array = [String]()
        for menuElement in subMenuList {
            for command in menuElement.getOwnCommandList() {
                array.append(command)
            }
        }
        return array
    }
    
    public func getSerialCommand() -> String {
        return serialCommand
    }
    
    public func getReturnCommandList() -> Array<String> {
        return returnCommandList
    }
    
    public func isTopic() -> Bool {
        return topic
    }
    
    public func isPause() -> Bool {
        if getSerialCommand()=="0" {
            return true
        } else {
            return false
        }
        
    }
    
    public func setName(name: String) {
        self.name=name
    }
    
    public func setAudioFilePath(path: String){
        audioFilePath=path
    }
    
    public func setCommandList(list: Array<String>){
        commandList=list
    }
    
    public func addCommand(command: String){
        commandList.append(command)
    }
    
    public func addSubMenuElement(element: GleimMenu){
        subMenuList.append(element)
    }
    
    public func setSerialCommand(serialCommand: String) {
        self.serialCommand=serialCommand
    }
    
    public func setReturnCommandList(returnCommandList: Array<String>) {
        self.returnCommandList=returnCommandList
    }
    
    public func addReturnCommand(command: String){
        returnCommandList.append(command)
    }
    public func setIsTopic(topic: Bool){
        self.topic = topic
    }
    
}

File: gleim300Code/JavaBackend/MistyBackend/src/model/GleimMenu.java
package model;

import javax.swing.tree.DefaultMutableTreeNode;
import java.util.ArrayList;

public class GleimMenu extends DefaultMutableTreeNode {

    private ArrayList<GleimMenu> subMenuList = new ArrayList<>();
    private String name = "";
    private String audioFilePath = "";
    private ArrayList<String> voiceCommandList = new ArrayList<>();
    private String serialCommand = "";
    private ArrayList<String> returnCommandList = new ArrayList<>();

    public GleimMenu(String name){
        super.setUserObject(name);
        this.name=name;
    }

    public ArrayList<String> getSubMenuCommandList(){
        ArrayList<String> array = new ArrayList<>();
        for(GleimMenu menuElement : subMenuList){
            for(String command : menuElement.getVoiceCommandList()){
                array.add(command);
            }
        }
        return  array;
    }

    public void addSubMenu(GleimMenu gleimMenu){
        super.add(gleimMenu);
        subMenuList.add(gleimMenu);
    }

    public void addVoiceCommand(String command){
        voiceCommandList.add(command);
    }

    public void addReturnCommand(String command){
        returnCommandList.add(command);
    }

    public void removeSubMenu(GleimMenu toRemove){
        super.remove(toRemove);
        subMenuList.remove(toRemove);
    }

    //-----------Generated Getter Setter------------

    public ArrayList<GleimMenu> getSubMenuList() {
        return subMenuList;
    }

    public void setSubMenuList(ArrayList<GleimMenu> subMenuList) {
        this.subMenuList = subMenuList;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        super.setUserObject(name);
        this.name = name;
    }

    public String getAudioFilePath() {
        return audioFilePath;
    }

    public void setAudioFilePath(String audioFilePath) {
        this.audioFilePath = audioFilePath;
    }

    public ArrayList<String> getVoiceCommandList() {
        return voiceCommandList;
    }

    public void setVoiceCommandList(ArrayList<String> voiceCommandList) {
        this.voiceCommandList = voiceCommandList;
    }

    public String getSerialCommand() {
        return serialCommand;
    }

    public void setSerialCommand(String serialCommand) {
        this.serialCommand = serialCommand;
    }

    public ArrayList<String> getReturnCommandList() {
        return returnCommandList;
    }

    public void setReturnCommandList(ArrayList<String> returnCommandList) {
        this.returnCommandList = returnCommandList;
    }

}
